File: TaskManager/labFinal_164/service/CommandManager.java
package labFinal_164.service;


import labFinal_164.command.ICommand;
import labFinal_164.repository.IssueRepository;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

public class CommandManager implements  ISubject {
    private Stack<ICommand> undoStack = new Stack<>();
    private Stack<ICommand> redoStack = new Stack<>();

    private List<IObserver> observers = new ArrayList<>();
    private IssueRepository repository;

    public CommandManager(IssueRepository repository) { this.repository = repository; }
    public void executeCommand(ICommand command) {
        command.execute();
        undoStack.push(command);
        redoStack.clear();
        notifyObservers(command);
    }
    public void undo() {
        if (!undoStack.isEmpty()) {
            ICommand command = undoStack.pop();
            command.undo();
            redoStack.push(command);
        }
    }
    public void redo() {
        if (!redoStack.isEmpty()) {
            ICommand command = redoStack.pop();
            command.execute();
            undoStack.push(command);
            notifyObservers(command);
        }
    }


    @Override
    public void registerObserver(IObserver observer) { observers.add(observer); }
    @Override
    public void removeObserver(IObserver observer) { observers.remove(observer); }
    @Override
    public void notifyObservers(ICommand command) {
        for (IObserver observer : observers) {
            observer.update(command);
        }
    }







}

File: TaskManager/TaskManager/Subjects/CommandManager.cs
using TaskManager.Commands;
using TaskManager.Observers;

namespace TaskManager.Subjects;

public class CommandManager : ISubject
{
    private readonly List<IObserver> _observers = [];
    private readonly Stack<ICommand> _executedCommands = new();
    private readonly Stack<ICommand> _undoneCommands = new ();

    public void RegisterObserver(IObserver observer)
    {
        if (!_observers.Contains(observer))
            _observers.Add(observer);
    }

    public void RemoveObserver(IObserver observer)
    {
        _observers.Remove(observer);
    }

    public void NotifyObservers(ICommand command)
    {
        foreach (var observer in _observers)
        {
            observer.Update(command);
        }
    }

    public void ExecuteCommand(ICommand command)
    {
        command.Execute();
        _executedCommands.Push(command);
        _undoneCommands.Clear();
        NotifyObservers(command);
    }

    public bool CanUndo() => _executedCommands.Count > 0;
    public bool CanRedo() => _undoneCommands.Count > 0;

    public void Undo()
    {
        if (!CanUndo()) return;

        var command = _executedCommands.Pop();
        command.Undo();
        _undoneCommands.Push(command);
        NotifyObservers(command);
    }

    public void Redo()
    {
        if (!CanRedo()) return;

        var command = _undoneCommands.Pop();
        command.Execute();
        _executedCommands.Push(command);
        NotifyObservers(command);
    }

    public IEnumerable<ICommand> GetCommandHistory()
    {
        return _executedCommands.Reverse();
    }
}